base/utf.jl

```diff
@@ -0,0 +1,880 @@
+
+# Get the trailing surrogate character in UTF-8 from an array, update the position
+macro get_utf8_surr!(str, pos)  ; quote ; ($(esc(pos)) += 3 ; ((UInt32($(esc(str))[$(esc(pos))-2] & 0xf) << 12) | (UInt32($(esc(str))[$(esc(pos))-1] & 0x3f) << 6) | ($(esc(str))[$(esc(pos))] & 0x3f))) ; end ; end
+
+# Output a character as UTF-8, update the position
+macro output_utf8_2!(buf, out, ch) ; quote ; $(esc(buf))[$(esc(out)) += 1] = 0xc0 | ($(esc(ch)) >>> 6) ;  $(esc(buf))[$(esc(out)) += 1] = 0x80 | ($(esc(ch)) & 0x3f) ; end ; end
+macro output_utf8_3!(buf, out, ch) ; quote ; $(esc(buf))[$(esc(out)) += 1] = 0xe0 | (($(esc(ch)) >>> 12) & 0x3f) ; $(esc(buf))[$(esc(out)) += 1] = 0x80 | (($(esc(ch)) >>> 6) & 0x3f) ; $(esc(buf))[$(esc(out)) += 1] = 0x80 | ($(esc(ch)) & 0x3f) ; end ; end
+macro output_utf8_4!(buf, out, ch) ; quote ; $(esc(buf))[$(esc(out)) += 1] = 0xf0 | ($(esc(ch)) >>> 18) ; $(esc(buf))[$(esc(out)) += 1] = 0x80 | (($(esc(ch)) >>> 12) & 0x3f) ; $(esc(buf))[$(esc(out)) += 1] = 0x80 | (($(esc(ch)) >>> 6) & 0x3f) ; $(esc(buf))[$(esc(out)) += 1] = 0x80 | ($(esc(ch)) & 0x3f) ; end ; end
+
+# Output a surrogate pair, update the position
+macro output_utf16_surr!(buf, out, ch) ; quote ; $(esc(buf))[$(esc(out)) += 1] = UInt16(0xd7c0 + ($(esc(ch)) >>> 10)) ; $(esc(buf))[$(esc(out)) += 1] = UInt16(0xdc00 + ($(esc(ch)) & 0x3ff)) ; end ; end
+
+#=
+"""
+@brief      encodes an AbstractString using UTF-16 encoding
+
+@param[in]  ::Type{UTF16String}
+@param[in]  str::AbstractString
+
+@return     ::UTF16String
+@throws     ArgumentError
+"""
+=#
+function convert(::Type{UTF16String}, str::AbstractString)
+    cnts::UTF_String_Counts = check_string_abs(str)
+    buf = Vector{UInt16}(cnts.cntT+cnts.cnt4+1)
+    out = 0
+    @inbounds for ch in str
+        c = reinterpret(UInt32, ch)
+        if c < 0x10000
+            buf[out += 1] = UInt16(c)
+        elseif c <= 0x10ffff
+            @output_utf16_surr!(buf, out, c)
+        else
+            throw(ArgumentError("invalid Unicode character (0x$(hex(c)) > 0x10ffff)"))
+        end
+    end
+    @inbounds buf[out += 1] = 0 # NULL termination
+    UTF16String(buf)
+end
+
+#=
+"""
+@brief      Converts an AbstractString to UTF-32 encoding
+
+@param[in]  ::Type{UTF32String}
+@param[in]  str::AbstractString
+
+@return     ::UTF32String
+@throws     ArgumentError
+"""
+=#
+function convert(::Type{UTF32String}, str::AbstractString)
+    len = length(str)
+    buf = Vector{Char}(len+1)
+    out = 0
+    @inbounds for ch in str ; buf[out += 1] = ch ; end
+    buf[out += 1] = Char(0) # NULL termination
+    UTF32String(buf)
+end
+
+#=
+@doc """
+@brief      Converts a UTF-8 encoded string to UTF-16 encoding
+
+@param[in]  src::Vector{UInt8}
+
+@return     ::UTF16String
+@throws     ArgumentError
+""" ->
+=#
+function convert(::Type{UTF16String}, src::UTF8String)
+    str = src.data
```

Comment:
BTW, in all the style guides I've seen (and those I've written myself!), one of the rules was not to use single character variable names, except for loop variables i,j,k, and then only when they are only used locally in the for loop.